interfaces/apparmor/spec.go

```diff
@@ -90,6 +96,32 @@ func (spec *Specification) AddSnippet(snippet string) {
 	}
 }
 
+// AddDeduplicatedSnippet adds a new apparmor snippet to all applications and hooks using the interface.
+//
+// Certain combinations of snippets may be computationally expensive for
+// apparmor_parser in its de-duplication step. This function lets snapd
+// perform de-duplication of identical rules at the potential cost of a
+// somewhat more complex auditing process of the text of generated
+// apparmor profile. Identical mount rules should typically use this, but
+// this function can also be used to avoid repeated rules that inhibit
+// auditability.
+func (spec *Specification) AddDeduplicatedSnippet(snippet string) {
```

Comment:
I wonder if we could always do this, it seems strange that we don't de-dup by default. Maybe a comment why this is not the default?

Reply:
I'm wondering the same... afaict it seems like this could be done implicitly and by default.

Reply:
because of the losing the trail of which interface(s) contributed the snippet, it can probably be solved, but naive using it for everything is not enough per @jdstrand requirements

Reply:
We had this proposal a while ago but it was NACked because it would make auditing the profile harder.

Reply:
I'll also mention that snapd's de-duplication is primarily useful for auto-generated policy for mount rules and snap-update-ns. What we are doing here for uio is primarily for tidiness *within a particular interface's rules*. We don't have a lot of duplication across interfaces and exact matches (mount rules notwithstanding) are trivially removed by the parser so the loss in auditability outweighs the benefit of policy with no duplicate rules.